pkg/workloadslicing/workloadslicing.go

```diff
@@ -217,12 +202,25 @@ func EnsureWorkloadSlices(ctx context.Context, clnt client.Client, jobPodSets []
 		// This transient state typically occurs when a new slice has been created,
 		// and the old slice is pending deactivation. The system should resolve this
 		// by deactivating the old slice, after which processing will continue under "case #1".
-
-		// There is also an edge-case when the old slice was preempted/evicted by the scheduler
-		// to make the room for other (than new slice) workload, which would result in two
-		// "pending" workloads. In such case - it is safe to deactivate the old slice.
 		oldWorkload := workloads[0]
-		if !workload.HasQuotaReservation(&oldWorkload) {
+
+		// Special case handling for an old workload slice.
+		//
+		// If the old workload has been evicted:
+		//   - If it still holds a quota reservation, finalize the eviction by clearing the
+		//     admission status and quota reservation.
+		//   - Then, mark the workload as finished and continue processing the new slice.
+		//
+		// If the workload is *not* evicted:
+		//   - If it still holds a quota reservation, it is assumed to be the active (current)
+		//     workload and should be left alone, it will be replaced by the scheduler.
+		//   - Otherwise, if it no longer has a quota reservation, finish it as it is out of sync.
+		if evictedCondition := apimeta.FindStatusCondition(oldWorkload.Status.Conditions, kueue.WorkloadEvicted); !workload.HasQuotaReservation(&oldWorkload) || evictedCondition != nil {
+			if workload.HasQuotaReservation(&oldWorkload) {
+				// Finalize eviction by resetting quota reservation and admission status.
+				_ = workload.UnsetQuotaReservationWithCondition(&oldWorkload, "Preempted", evictedCondition.Message, time.Now())
```

Comment:
Are these changes needed? I'm surpised becuase marking workload as Finished should automatically exclude it from reserving quota, even if QuotaReserverd=True. Note that when we normally finish the Job we just mark the workload as finished: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L439-L451

Please verify if the changes are needed, and revert if not needed. 

If they are needed, this is not just a cleanup but bugfix, and I think we should reflect that in the PR status + cherrypick.

Reply:
> Are these changes needed? I'm surpised becuase marking workload as Finished should automatically exclude it from reserving quota

Good call out. You are correct, we don't need to "workload.UnsetQuotaReservationWithCondition" and "Finish"-ing works as intended. 

Removing.